calico/openstack/mech_calico.py

```diff
@@ -120,146 +173,437 @@ def update_subnet_postcommit(self, context):
+            self.add_port_gateways(port, context._plugin_context)
+            self.add_port_interface_name(port)
+            port['security_groups'] = self.get_security_groups_for_port(
+                context._plugin_context, port
+            )
+            profiles = self.get_security_profiles(
+                context._plugin_context, port
+            )
+            self.transport.endpoint_created(port)
 
-        # Get a DB context for this query.
-        db_context = ctx.get_admin_context()
+            for profile in profiles:
+                self.transport.write_profile_to_etcd(profile)
 
-        # Get current SGs.
-        sgs = self.db.get_security_groups(db_context)
+    def add_port_gateways(self, port, context):
+        """
+        Determine the gateway IP addresses for a given port's IP addresses, and
+        adds them to the port dict.
+
+        This method assumes it's being called from within a database
+        transaction and does not take out another one.
+        """
+        for ip in port['fixed_ips']:
+            subnet = self.db.get_subnet(context, ip['subnet_id'])
+            ip['gateway'] = subnet['gateway_ip']
 
-        # Add, to each SG, a dict whose keys are the endpoints configured to
-        # use that SG, and whose values are the corresponding IP addresses.
-        for sg in sgs:
-            sg['members'] = self._get_members(sg, db_context)
+    def get_security_profiles(self, context, port):
+        """
+        Obtain information about the security profile that applies to a given
+        port.
+
+        This method expects to be called from within a database transaction,
+        and does not create its own.
+
+        :returns: A generator of ``SecurityProfile`` objects.
+        """
+        # For each security group get its rules. Given that we don't need
+        # anything else about the security group, we can do this as a single
+        # query.
+        # CB2: I am concerned that this does not adequately prevent new
+        # security group rules being added and racing us in.
+        sgids = port['security_groups']
+        rules = self.db.get_security_group_rules(
+            context, filters={'security_group_id': sgids}
+        )
+
+        # Now, return a generator that provides profile objects for each
+        # profile.
+        return (
+            profile_from_neutron_rules(sgid, rules) for sgid in sgids
+        )
+
+    def periodic_resync_thread(self):
+        """
+        This method acts as a the periodic resynchronization logic for the
+        Calico mechanism driver.
+
+        On a fixed interval, it spins over the entire database and reconciles
+        it with etcd, ensuring that the etcd database and Neutron are in
+        synchronization with each other.
+        """
+        # The very first thing we do is sleep for our startup interval before
+        # initializing our state.
+        eventlet.sleep(STARTUP_DELAY_SECS)
```

Comment:
I wasn't clear enough here: I'd like your input on whether we can abandon start of day sync. For the moment I've concluded we can't, so this is the approach that will stay in for the rest of today. If you think we can give up on start of day, let's re-evaluate. Feel free to call me later today to talk it over.